src/OpenApi/src/Services/OpenApiDocumentService.cs

```diff
@@ -19,12 +20,18 @@ internal sealed class OpenApiDocumentService(
 {
     private readonly OpenApiOptions _options = optionsMonitor.Get(documentName);
 
+    // For good hygiene, operation-level tags must also appear in the document-level
+    // tags collection. This set captures all tags that have been seen so far.
+    // Note: internal for testing.
+    internal readonly HashSet<OpenApiTag> _capturedTags = new(OpenApiTagComparer.Instance);
+
     public Task<OpenApiDocument> GetOpenApiDocumentAsync()
     {
         var document = new OpenApiDocument
         {
             Info = GetOpenApiInfo(),
-            Paths = GetOpenApiPaths()
+            Paths = GetOpenApiPaths(),
```

Comment:
This is basically just a style preference, but I would find it more readable if the set were local to `GetOpenApiDocumentAsync` and threaded through as a parameter.  That would make it clear, for example, that there are no thread safety considerations and that the set is recomputed (idempotently, at the moment) on each call to `GetOpenApiDocumentAsync` (maybe there's only one?).

Reply:
> (idempotently, at the moment) ... (maybe there's only one?).

Good note! Yes, the tags are recomputed idempotently on each `GetOpenApiDocumentAsync` invocation. `GetOpenApiDocumentAsync` is currently called every time the user hits the `/openapi/{documentName}.json` endpoint. It is a little harder to observe this the way it's currently structured so I'll adjust.